projects/update_success_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,344 @@
+
+## Make sure that the volume of incoming pings is reasonable
+Check if the volume of `update` pings with `reason = ready` matches with the volume of pings with `reason = success`. For each ping with `reason = ready`, find the matching ping with `reason = success`
+
+
+```python
+# Get an identifier that keeps in consideration both the current build
+# and the target build.
+ready_uuid = ping_ready\
+    .map(lambda p: (p.get("clientId"),
+                    p.get("application/buildId"),
+                    p.get("application/channel"),
+                    p.get("application/version"),
+                    p.get("payload/targetBuildId"),
+                    p.get("payload/targetChannel"),
+                    p.get("payload/targetVersion")))
+
+# Get an identifier that considers both the prevous build info and the
+# current build info. The order of the values in the tuple need to match
+# the one from the previous RDD.
+success_uuid = ping_success\
+    .map(lambda p: (p.get("clientId"),
+                    p.get("payload/previousBuildId"),
+                    p.get("payload/previousChannel"),
+                    p.get("payload/previousVersion"),
+                    p.get("application/buildId"),
+                    p.get("application/channel"),
+                    p.get("application/version")))
+```
+Let's match each `reason = ready` ping with a `reason = success` one, and count them.
+
+
+```python
+matching_update_pings = ready_uuid.intersection(success_uuid)
+matching_update_ping_count = matching_update_pings.count()
+```
+Finally, show up some stats.
+
+
+```python
+print("{:.3f}% of the 'update' ping with reason 'ready' have a matching ping with reason 'success'."\
+      .format(pct(matching_update_ping_count, ping_ready_count)))
+print("{:.3f}% of the 'update' ping with reason 'success' have a matching ping with reason 'ready'."\
+      .format(pct(matching_update_ping_count, ping_success_count)))
+```
+    63.834% of the 'update' ping with reason 'ready' have a matching ping with reason 'success'.
+    79.908% of the 'update' ping with reason 'success' have a matching ping with reason 'ready'.
+
+
+Only ~63% of the update ping sent when an update is ready to be applied have a corrensponding ping that's sent, for the same client and upgrade path, after the update is successfully applied. One possible explaination for this is the delay with which updates get applied after they get downloaded: unless the browser is restarted (and that can happen after days due to user suspending their machines), we won't see the ready ping anytime soon. 
+
+Roughly 80% of the `update` pings with `reason = success` can be traced back to an `update` with `reason = ready`. The missing ~20% matches can be due to users disabling automatic updates: no `update` ping is sent [in that case](https://bugzilla.mozilla.org/show_bug.cgi?id=1386619) if an update is manually triggered.
```

Comment:
File a bug to check the %ge of this population that have automatic updates turned off?

Also, this analysis is a little flawed on the edges. For instance, someone could be updating to 20170903 from 20170901 so the "ready" ping will be missing but the "success" ping will be present. Similarly for someone updating from 20170909 to 20170911. "ready" will be present, but "success" will be missing.

Could fix this by excluding any "success" coming from a previousBuildId behind the window and any "ready" pings going to a targetBuild that's after the window.

Reply:
Good point. I've smoothed the edges!